optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -554,4 +561,171 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
     private ConstraintCollectors() {
     }
 
+    // ************************************************************************
+    // toCollection
+    // ************************************************************************
+
+    /**
+     * Creates constraint collector that will return matched facts as a collection.
+     *
+     * @param collectionFunction never null, provides new instance of the target collection with given initial capacity
+     * @param <A> type of the matched fact, matching the type of elements in the resulting collection
+     * @param <Result> type of the target collection
+     * @return never null
+     */
+    public static <A, Result extends Collection<A>> UniConstraintCollector<A, ?, Result> toCollection(
+            IntFunction<Result> collectionFunction) {
+        return toCollection(Function.identity(), collectionFunction);
+    }
+
+    public static <A> UniConstraintCollector<A, ?, Set<A>> toSet() {
+        return toSet(Function.identity());
+    }
+
+    public static <A> UniConstraintCollector<A, ?, List<A>> toList() {
+        return toList(Function.identity());
+    }
+
+    /**
+     * Creates constraint collector that will return a collection of the given type.
+     *
+     * @param mappingFunction never null, converts matched facts to elements of the resulting collection
+     * @param collectionFunction never null, provides new instance of the target collection with given initial capacity
+     * @param <A> type of the matched fact
+     * @param <Mapped> type of elements in the resulting collection
+     * @param <Result> type of the resulting collection
+     * @return never null
+     */
+    public static <A, Mapped, Result extends Collection<Mapped>> UniConstraintCollector<A, ?, Result> toCollection(
+            Function<A, Mapped> mappingFunction, IntFunction<Result> collectionFunction) {
+        return new DefaultUniConstraintCollector<>(
+                (Supplier<List<Mapped>>) ArrayList::new,
+                (resultContainer, a) -> {
+                    Mapped mapped = mappingFunction.apply(a);
+                    resultContainer.add(mapped);
+                    return () -> resultContainer.remove(mapped);
+                },
+                resultContainer -> finishToCollection(collectionFunction, resultContainer));
+    }
+
+    private static <Mapped, Container extends List<Mapped>, Result extends Collection<Mapped>> Result
+    finishToCollection(IntFunction<Result> collectionFunction, Container resultContainer) {
```

Comment:
Thinking aloud: The method name doesn't indicate it's used as the undo operation.
How about calling it to undoToCollection?

Reply:
This is not the undo. This is the finisher, which takes the incremental container and turns it into a final immutable product.

Reply:
Agreed, then the name is correct. I interpreted "finish" as last part that is common, not as finisher.

Maybe `toCollectionFinisher`? Although name is fine now.